File: Editorial/55_Jump_Game/Go/3_Dynamic_Programming_Bottom_up.go
func canJump(nums []int) bool {
    type Index int
    const (
        GOOD Index = iota
        BAD
        UNKNOWN
    )
    memo := make([]Index, len(nums))
    for i := range memo {
        memo[i] = UNKNOWN
    }
    memo[len(memo)-1] = GOOD
    for i := len(nums) - 2; i >= 0; i-- {
        furthestJump := min(i+nums[i], len(nums)-1)
        for j := i + 1; j <= furthestJump; j++ {
            if memo[j] == GOOD {
                memo[i] = GOOD
                break
            }
        }
    }
    return memo[0] == GOOD
}

func min(a, b int) int {
    if a < b {
        return a
    }
    return b
}

File: Editorial/0000-0100/0055_Jump_Game/TypeScript/3_Dynamic_Programming_Bottom_up.ts
enum Index {
    GOOD,
    BAD,
    UNKNOWN,
}
function canJump(nums: number[]): boolean {
    let memo: Index[] = Array(nums.length).fill(Index.UNKNOWN);
    memo[memo.length - 1] = Index.GOOD;
    for (let i = nums.length - 2; i >= 0; i--) {
        let furthestJump: number = Math.min(i + nums[i], nums.length - 1);
        for (let j = i + 1; j <= furthestJump; j++) {
            if (memo[j] == Index.GOOD) {
                memo[i] = Index.GOOD;
                break;
            }
        }
    }
    return memo[0] == Index.GOOD;
}
